Document/0x05d-Testing-Data-Storage.md

```diff
@@ -271,11 +271,104 @@ You can use stored keys in one of two modes:
+
+During key attestation, we can specify the alias of a key pair and in return, get a certificate chain, which we can use to verify the properties of that key pair. If the root certificate of the chain is [Google Hardware Attestation Root certificate](https://developer.android.com/training/articles/security-key-attestation#root_certificate "Google Hardware Attestation Root certificate") and the checks related to key pair storage in hardware are made it gives an assurance that the device supports hardware-level key attestation and the key is in hardware-backed keystore that Google believes to be secure. Alternatively, if the attestation chain has any other root certificate, then Google does not make any claims about the security of the hardware.
+
+Although the key attestation process can be implemented within the application directly but it is recommended that it should be implemented at the server-side for security reasons. The following are the high-level guidelines for the secure implementation of Key Attestation:
+
+- The server should initiate the key attestation process by creating a random number securely using CSPRNG(Cryptographically Secure Random Number Generator) and the same should be sent to the user as a challenge.
+- The client should call the `setAttestationChallenge` API with the challenge received from the server and should then retrieve the attestation certificate chain using the `KeyStore.getCertificateChain` method.
+- The attestation response should be sent to the server for the verification and following checks should be performed for the verification of the key attestation response:
+  - Verify the certificate chain, up to the root and perform certificate sanity checks such as validity, integrity and trustworthiness.
+  - Check if the root certificate is signed with the Google attestation root key which makes the attestation process trustworthy.
+  - Extract the attestation certificate extension data, which appears within the first element of the certificate chain and perform the following checks:
+    - Verify that the attestation challenge is having the same value which was generated at the server while initiating the attestation process.
+    - Verify the signature in the key attestation response.
+    - Now check the security level of the Keymaster to determine if the device has secure key storage mechanism. Keymaster is a piece of software that runs in the security context and provides all the secure keystore operations. The security level will be one of `Software`, `TrustedEnvironment` or `StrongBox`.
```

Comment:
note that software version can be done by any... so be careful then. take TE and SB as a real security hook, or you will need to re-evaluate the threatmodel in place as it is one big compromisable software party.

Reply:
Not sure what you had in mind here. In Key Attestation section there is no `software version` keyword and I'm not able to find relation.

Reply:
ah no, i mean security level being "software". in that case it might be faked by the device and/or an attacker. there were some notes / warnings 👍 in https://developer.android.com/training/articles/security-key-attestation as well:

```Most likely, the device launched with an Android version less than 7.0 and it does not support hardware attestation. In this case, Android has a software implementation of attestation which produces the same sort of attestation certificate, but signed with a key hardcoded in Android source code. Because this signing key is not a secret, the attestation could have been created by an attacker pretending to provide secure hardware.```

Reply:
That is clear :) I tried clarifying this - device which support hardware-key attestation and what to check.